File: backup/Twitch/app/src/main/java/com/exact/twitch/ui/player/BasePlayerFragment.java
package com.exact.twitch.ui.player;

import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.content.res.Configuration;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.View;
import android.widget.ImageButton;

import com.exact.twitch.R;
import com.exact.twitch.di.Injectable;
import com.exact.twitch.ui.fragment.OnChannelClickedListener;
import com.exact.twitch.ui.view.SlidingView;
import com.exact.twitch.ui.view.draggableview.DraggableListener;
import com.exact.twitch.ui.view.draggableview.DraggableView;
import com.google.android.exoplayer2.ui.PlayerView;

import javax.inject.Inject;

public abstract class BasePlayerFragment extends Fragment implements Injectable {

    protected final String TAG = getClass().getSimpleName();
    protected PlayerView pvPlayer;
    protected DraggableView draggableView;
    protected SlidingView slidingView;
    protected ImageButton channelBtn;
    protected ImageButton moreBtn;
    protected OnChannelClickedListener channelListener;

    private DraggableListener dragListener;
    private boolean isPortraitOrientation;

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        isPortraitOrientation = getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        pvPlayer = view.findViewById(R.id.fragment_player_pv_player);
        channelBtn = view.findViewById(R.id.player_profile);
        moreBtn = view.findViewById(R.id.player_more);
        ImageButton minimizeBtn = view.findViewById(R.id.player_minimize);
        minimizeBtn.setOnClickListener(v -> minimize());
        ImageButton fullscreenBtn;
        if (isPortraitOrientation) {
            draggableView = view.findViewById(R.id.fragment_player_draggable_view);
            draggableView.setDraggableListener(dragListener);
            fullscreenBtn = view.findViewById(R.id.player_fullscreen);
            fullscreenBtn.setOnClickListener(v -> requireActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE));
        } else {
//            slidingView = view.findViewById(R.id.fragment_player_sv);
            fullscreenBtn = view.findViewById(R.id.player_fullscreen_exit);
            fullscreenBtn.setOnClickListener(v -> requireActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT));
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        if (!requireActivity().isChangingConfigurations()) {
            pvPlayer.setPlayer(null);
        }
    }

    public void minimize() {
        if (isPortraitOrientation) {
            draggableView.minimize();
        } else {
//            slidingView.minimize();
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnChannelClickedListener) {
            channelListener = (OnChannelClickedListener) context;
        } else {
            throw new RuntimeException(context.toString() + " must implement OnViewChannelClickedListener");
        }
        if (context instanceof DraggableListener) {
            dragListener = (DraggableListener) context;
        } else {
            throw new RuntimeException(context.toString() + " must implement DraggableListener");
        }
    }
}

File: app/src/main/java/com/exact/xtra/ui/player/BasePlayerFragment.kt
package com.exact.xtra.ui.player

import android.content.Context
import android.content.pm.ActivityInfo
import android.content.res.Configuration
import android.os.Bundle
import android.os.Parcelable
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.ViewModelProvider
import com.exact.xtra.di.Injectable
import com.exact.xtra.ui.common.OnChannelClickedListener
import com.exact.xtra.ui.view.draggableview.DraggableListener
import com.exact.xtra.ui.view.draggableview.DraggableView
import com.google.android.exoplayer2.Player
import kotlinx.android.synthetic.main.player_stream.*
import javax.inject.Inject

@Suppress("PLUGIN_WARNING")
abstract class BasePlayerFragment : Fragment(), Injectable, LifecycleObserver {

    private var channelListener: OnChannelClickedListener? = null
    private var dragListener: DraggableListener? = null
    private var isPortraitOrientation: Boolean = false
    private lateinit var draggableView: DraggableView
    protected abstract val viewModel: PlayerViewModel

    @Inject
    protected lateinit var viewModelFactory: ViewModelProvider.Factory

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnChannelClickedListener) {
            channelListener = context
        } else {
            throw RuntimeException(context.toString() + " must implement OnViewChannelClickedListener")
        }
        if (context is DraggableListener) {
            dragListener = context
        } else {
            throw RuntimeException(context.toString() + " must implement DraggableListener")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isPortraitOrientation = resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT

    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (isPortraitOrientation) {
            minimize.setOnClickListener { minimize() }
            draggableView = view as DraggableView
            draggableView.setDraggableListener(dragListener)
            fullscreenEnter.setOnClickListener { requireActivity().requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE }
        } else {
            //            slidingView = view.findViewById(R.id.fragment_player_sv);
            fullscreenExit.setOnClickListener { requireActivity().requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT }
        }
    }

    override fun onResume() {
        super.onResume()
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
    }

    override fun onDetach() {
        super.onDetach()
        channelListener = null
        dragListener = null
    }

    abstract fun play(obj: Parcelable)

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    open fun onMoveToForeground() {
        println("resume")
        if (viewModel.player.playbackState != Player.STATE_READY) {
            viewModel.startPlayer()
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    open fun onMoveToBackground() {
        println("stop")
        viewModel.player.stop()
    }

    fun minimize() {
        if (isPortraitOrientation) {
            draggableView.minimize()
        } else {
//                        slidingView.minimize();
        }
    }
}
